10_cdev/cdev.c

```diff
@@ -0,0 +1,424 @@
+static int cdev_release(struct inode *inode, struct file *file)
+{
+	files_counter--;
+
+	return add_note_to_list("File was closed");
+}
+
+/**
+ * cdev_read() - called on file read() operation
+ * @file: VFS file opened by a process
+ * @buf:
+ * @count:
+ * @loff:
+ */
+static ssize_t cdev_read(struct file *file, char __user *buf, size_t count,
+			 loff_t *loff)
+{
+	int status = add_note_to_list("File was read");
+
+	if (IS_ERR_VALUE(status))
+		return status;
+
+	if (used_buf_vol == 0)
+		pr_info("Buffer is empty:(\n");
+
+	return simple_read_from_buffer(buf, count, loff, buffer,
+				       strlen(buffer));
+}
+
+/**
+ * cdev_write() - callback for file write() operation
+ * @file: VFS file opened by a process
+ * @buf:
+ * @count:
+ * @loff:
+ */
+static ssize_t cdev_write(struct file *file, const char __user *buf,
+			  size_t count, loff_t *loff)
+{
+	uint32_t n = 0;
+	int status = 0;
+	char *tmp_note;
+	char *tmp = kzalloc(sizeof(*tmp) * count, GFP_KERNEL);
+
+	if (tmp == NULL) {
+		pr_err("Can't write to buffer:(\n");
+		return 0;
+	}
+
+	n = simple_write_to_buffer(tmp, count, loff, buf, count);
+
+	spin_lock(&lock);
+
+	if (count > buffer_size) {
+		status = add_note_to_list("**Input string too large");
+		if (IS_ERR_VALUE(status))
+			return status;
+	}
+
+	if ((strlen(buffer) + count) >= buffer_size) {
+		strncpy(buffer, tmp, buffer_size);
+		status = add_note_to_list("**Buffer was overrided");
+		if (IS_ERR_VALUE(status))
+			return status;
+
+	} else {
+		strcat(buffer, "\n");
+		strcat(buffer, tmp);
+	}
+
+	used_buf_vol = strlen(buffer);
+
+	spin_unlock(&lock);
+
+	tmp_note = kzalloc(sizeof(*tmp) * (strlen(tmp) +
+					   strlen("'' was written to buffer")),
+			   GFP_KERNEL);
+	if (tmp_note == NULL) {
+		pr_err("Can't write to journal:(\n");
+		return 0;
+	}
+
+	sprintf(tmp_note, "'%s' was written to buffer",
+		count > buffer_size ? buffer : tmp);
+
+	status = add_note_to_list(tmp_note);
+	if (IS_ERR_VALUE(status))
+		return status;
+
+	kfree(tmp);
+	kfree(tmp_note);
+
+	return n;
+}
+
+// This structure is partially initialized here
+// and the rest is initialized by the kernel after call
+// to cdev_init()
+static const struct file_operations cdev_fops = {
+	.open = &cdev_open,
+	.release = &cdev_release,
+	.read = &cdev_read,
+	.write = &cdev_write,
+	// required to prevent module unloading while fops are in use
+	.owner = THIS_MODULE,
+};
+
+/* sysfs show() method. Calls the show() method corresponding to the individual sysfs file */
+static ssize_t used_buf_vol_show(struct class *class,
+				 struct class_attribute *attr, char *buf)
+{
+	sprintf(buf, "%d\n", used_buf_vol);
+
+	return strlen(buf);
+}
+CLASS_ATTR_CREATE(used_buf_vol, 0444, &used_buf_vol_show, NULL);
+
+/* sysfs show() method. Calls the show() method corresponding to the individual sysfs file */
+static ssize_t buffer_size_show(struct class *class,
+				struct class_attribute *attr, char *buf)
+{
+	sprintf(buf, "%d\n", buffer_size);
+
+	return strlen(buf);
```

Comment:
just for note: sprintf return the length of result string. so strlen call can be omitted.